### PR TITLE
KokkosBatched - QR and its test

### PR DESCRIPTION
**Pull request by kyungjoo-kim**
Batch QR is requested from other applications (intrepid2 and compadre). This PR includes QR infrastructure for team level functions.

So far the batched dense linear algebra routines are not tested in the cmake environment. This also turns on the tests. This also passes cm test all on kokkos-dev-2.

### CODE REVIEW

**Comment by kyungjoo-kim** [comment]
White... 
```
[kyukim@white27 test-all]$  ../../../kyungjoo/scripts/cm_test_all_sandia gcc/6.4.0 cuda/10.1.105
Running on machine: white

KokkosKernels Repository Status:  0112f3c2f58bb26e9633638c94ef58c0d703bdea KokkosBatched - QR and its test

Kokkos Repository Status:  0a633a82c8c8c453f406c8d6356c6dfad64307bb Merge pull request #2628 from ibaned/cuda-10-defaulted


Going to test compilers:  gcc/6.4.0 cuda/10.1.105
Testing compiler gcc/6.4.0
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
  Starting job gcc-6.4.0-OpenMP-release
kokkos options: 
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-6.4.0-OpenMP-release
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
  Starting job gcc-6.4.0-Serial-release
kokkos options: 
kokkos devices: Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-6.4.0-Serial-release
Testing compiler cuda/10.1.105
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
  Starting job gcc-6.4.0-OpenMP_Serial-release
kokkos options: 
kokkos devices: OpenMP,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-6.4.0-OpenMP_Serial-release
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
  Starting job cuda-10.1.105-Cuda_OpenMP-release
kokkos options: 
kokkos devices: Cuda,OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED cuda-10.1.105-Cuda_OpenMP-release
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
  Starting job cuda-10.1.105-Cuda_Serial-release
kokkos options: 
kokkos devices: Cuda,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED cuda-10.1.105-Cuda_Serial-release
#######################################################
PASSED TESTS
#######################################################
cuda-10.1.105-Cuda_OpenMP-release build_time=365 run_time=191
cuda-10.1.105-Cuda_Serial-release build_time=370 run_time=232
gcc-6.4.0-OpenMP-release build_time=105 run_time=47
gcc-6.4.0-OpenMP_Serial-release build_time=134 run_time=136
gcc-6.4.0-Serial-release build_time=101 run_time=84
```

File: src/batched/KokkosBatched_ApplyQ_Serial_Impl.hpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+#include "KokkosBatched_Util.hpp"
+#include "KokkosBatched_ApplyQ_Serial_Internal.hpp"
+
+namespace KokkosBatched {
+
+  ///
+  /// Serial Impl
+  /// ===========
+
+  template<>
```
Comment (ndellingwood) [inline src/batched/KokkosBatched_ApplyQ_Serial_Impl.hpp:16]: Why is there the empty `template<>` followed by another `template<...>`?

File: src/batched/KokkosBatched_ApplyQ_Serial_Impl.hpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+         const wViewType &w) {
+    return SerialApplyQ_LeftForwardInternal::
+      invoke(B.extent(0), B.extent(1), A.extent(1), 
+             A.data(), A.stride_0(), A.stride_1(),
+             t.data(), t.stride_0(), 
+             B.data(), B.stride_0(), B.stride_1(),
+             w.data());
+  }
+
+  template<>
```
Comment (ndellingwood) [inline src/batched/KokkosBatched_ApplyQ_Serial_Impl.hpp:36]: Why is there the empty `template<>` followed by another `template<...>`?

File: src/batched/KokkosBatched_ApplyQ_Serial_Impl.hpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+         const wViewType &w) {
+    return SerialApplyQ_LeftBackwardInternal::
+      invoke(B.extent(0), B.extent(1), A.extent(1), 
+             A.data(), A.stride_0(), A.stride_1(),
+             t.data(), t.stride_0(), 
+             B.data(), B.stride_0(), B.stride_1(),
+             w.data());
+  }
+
+  template<>
```
Comment (ndellingwood) [inline src/batched/KokkosBatched_ApplyQ_Serial_Impl.hpp:56]: Why is there the empty `template<>` followed by another `template<...>`?

File: unit_test/CMakeLists.txt [hunk 4]
```diff
@@ -11,8 +11,9 @@ KOKKOSKERNELS_INCLUDE_DIRECTORIES(REQUIRED_DURING_INSTALLATION_TESTING ${CMAKE_C
 KOKKOSKERNELS_INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR}/graph)
 KOKKOSKERNELS_INCLUDE_DIRECTORIES(REQUIRED_DURING_INSTALLATION_TESTING ${CMAKE_CURRENT_SOURCE_DIR}/graph)
 
-#INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR}/batched)
-#INCLUDE_DIRECTORIES(REQUIRED_DURING_INSTALLATION_TESTING ${CMAKE_CURRENT_SOURCE_DIR}/batched)
+# Kyungjoo : somehow, this is not included thus, it is never tested with cmake
```
Comment (ndellingwood) [inline unit_test/CMakeLists.txt:14]: Should comment be removed?

File: src/batched/KokkosBatched_ApplyQ_Serial_Impl.hpp (hunk 1 above)
Comment (kyungjoo-kim) [reply src/batched/KokkosBatched_ApplyQ_Serial_Impl.hpp:16]: It is partial specialization. The first template argument is struct template allowing partial specialization and the followed template arguments are function templates.

File: src/batched/KokkosBatched_ApplyQ_Serial_Impl.hpp (hunk 2 above)
Comment (kyungjoo-kim) [reply src/batched/KokkosBatched_ApplyQ_Serial_Impl.hpp:36]: This struct is specialized by its struct template arguments Side::Left, Trans::Trnaspose, Algo::ApplyQ::Unblocked. That is why it is empty but the struct has specialization while the function still can get views as a free type.

File: src/batched/KokkosBatched_ApplyQ_Serial_Impl.hpp (hunk 3 above)
Comment (kyungjoo-kim) [reply src/batched/KokkosBatched_ApplyQ_Serial_Impl.hpp:56]: Same as above.

File: unit_test/CMakeLists.txt (hunk 4 above)
Comment (kyungjoo-kim) [reply unit_test/CMakeLists.txt:14]: I can remove the comment.

Comment (kyungjoo-kim) [reply unit_test/CMakeLists.txt]: Done.

**Comment by ndellingwood** [comment]
@kyungjoo-kim was waiting on the Kokkos+KokkosKernels 3.1 PR to merge into Trilinos, merging this now.

**Pull request merged**